### PR TITLE
ENH: Adds conditional VAE experiments

### PR DESCRIPTION
**Pull request by itellaetxe**
## Summary
PR to discuss results and potential improvements on the conditional VAE implementation based on [this paper](https://doi.org/10.1007/978-3-030-32245-8_91) and their [code](https://github.com/QingyuZhao/VAE-for-Regression/blob/master).

---
After implementing the conditional VAE as in the provided reference and training it for 64 epochs (Early stopped due to lack of improvement in the MSE) I get the following results. Note that I conditioned the network on the **length** of the streamlines, as this is a fairly easy thing to learn from the morphology of the streamlines.

- The losses: `kl_loss: 343.6720 - label_loss: 329.0857 - reconstruction_loss: 28.0201 - total_loss: 700.777`. This was epoch number 14. Very early.

- The reconstruction (very crappy, but it kind of resembles the FiberCup tractogram): 

- The latent space representation using t-SNE (`perplexity=30, early_exaggeration=12.0`): 

### Observations
- It does look like the bundles get encoded into individual and well defined clusters. Good news! The network understands morphological differences in the bundles and knows that they are "different things". The dataset has 7 distinct bundles, and we got 7 clusters. I should visualize the bundle label of the streamlines to see if what I am assuming here is correct, but I am pretty confident I am right.

- Looking at the morphology of the clusters, they stretch along dimension 2, so the conditioning might be working properly.

### Future work
- Improve the results of the training -> Training for longer, optimizing hyperparameters...
- Evaluate how good the latent space is regularized according to the conditioning variable (length in this case, should be age in the future.
- Apply to human tractograms.

### CODE REVIEW

**Comment by pjsjongsung** [comment]
Great work! I like that it is divided into 7 clusters. Is it possible that during early stopping mse was not improving but label loss was?

Also forgive me if we already talked about this, does the 7 clusters look very different from when you train without conditions? I'm just curious if the length is helping in clustering them or not.

And you said the clusters are stretching in dimension 2, which could mean the length is working. Is there a way to verify this? Like could we set a colormap and see if there is a gradual increase in color across dimension 2?

**Comment by jhlegarreta** [comment]
Thanks Iñigo.

>> It does look like the bundles get encoded into individual and well defined clusters. Good news! The network understands morphological differences in the bundles and knows that they are "different things". The dataset has 7 distinct bundles, and we got 7 clusters. I should visualize the bundle label of the streamlines to see if what I am assuming here is correct, but I am pretty confident I am right.

> Great work! I like that it is divided into 7 clusters. Is it possible that during early stopping mse was not improving but label loss was?
> Also forgive me if we already talked about this, does the 7 clusters look very different from when you train without conditions? I'm just curious if the length is helping in clustering them or not.

Agree with these comments. We would need to train for more epochs or find the hyperparameters to achieve a better reconstruction. Also, if we look at the FiberCup latent space in FINTA, the clusters were identifiable without conditioning the AE. Although I confess that this t-SNE projection looks better structured, the variational part may have played a role as well.

>> Looking at the morphology of the clusters, they stretch along dimension 2, so the conditioning might be working properly.

> And you said the clusters are stretching in dimension 2, which could mean the length is working. Is there a way to verify this? Like could we set a colormap and see if there is a gradual increase in color across dimension 2?

Coloring by the condition is necessary here, otherwise we are guessing. This should be easy to achieve with matplotlib.

Also, note that the latent space projection depends on the input data, so when the data that you are projecting changes, the appearance of the latent space will change. The relevant thing is to see whether there transition across lengths within each cluster is still kept. Note that in this case (and unless we add implausibles to the set), most lengths will be similar, so it may be a hard case to see those transitions within a cluster.

**Review by jhlegarreta — COMMENTED** [review]
Thanks Iñigo.

>> It does look like the bundles get encoded into individual and well defined clusters. Good news! The network understands morphological differences in the bundles and knows that they are "different things". The dataset has 7 distinct bundles, and we got 7 clusters. I should visualize the bundle label of the streamlines to see if what I am assuming here is correct, but I am pretty confident I am right.

> Great work! I like that it is divided into 7 clusters. Is it possible that during early stopping mse was not improving but label loss was?
> Also forgive me if we already talked about this, does the 7 clusters look very different from when you train without conditions? I'm just curious if the length is helping in clustering them or not.

Agree with these comments. We would need to train for more epochs or find the hyperparameters to achieve a better reconstruction. Also, if we look at the FiberCup latent space in FINTA, the clusters were identifiable without conditioning the AE. Although I confess that this t-SNE projection looks better structured, the variational part may have played a role as well.

>> Looking at the morphology of the clusters, they stretch along dimension 2, so the conditioning might be working properly.

> And you said the clusters are stretching in dimension 2, which could mean the length is working. Is there a way to verify this? Like could we set a colormap and see if there is a gradual increase in color across dimension 2?

Coloring by the condition is necessary here, otherwise we are guessing. This should be easy to achieve with matplotlib.

Also, note that the latent space projection depends on the input data, so when the data that you are projecting changes, the appearance of the latent space will change. The relevant thing is to see whether there transition across lengths within each cluster is still kept. Note that in this case (and unless we add implausibles to the set), most lengths will be similar, so it may be a hard case to see those transitions within a cluster.

**Comment by itellaetxe** [comment]
@pjsjongsung @jhlegarreta

I forgot to mention that these projections are only considering the **plausible** fibers. Bearing in mind that the losses of the model were quite high when it stopped training, I believe that the model did still not learn to distinguish (unsupervisedly) plausibles from implausibles. Anyways, here is the plot with both fiber types:

It may be better to plot this in 3D (third dimension being the length of the streamlines) and assign the coloring based on the bundle.

> Great work! I like that it is divided into 7 clusters. Is it possible that during early stopping mse was not improving but label loss was?

That was the case indeed. There was a complex interplay between the three components of the loss, so there were times when one loss was improving and the other were two getting worse, etc. But yes, mainly KL loss and label loss were improving. Maybe we could add a weighing parameter to the label loss to balance this a bit?

> Also forgive me if we already talked about this, does the 7 clusters look very different from when you train without conditions? I'm just curious if the length is helping in clustering them or not.

Yes, the clusters do look different when we train without conditioning. While @jhlegarreta was right saying that without coloring, we can't confirm whether the conditioning works, I think that visualization is not enough to truly know this. The t-SNE representation may have aligned other dimensions more meaningful to the algorithm, that do not necessarily include the conditioning variable (streamline length). Maybe the authors of the code I based mine were lucky to get the alignment, or smart enough to manipulate the conditioning variable (normalization, Z-scoring, etc.) to make t-SNE grab this information and put it along an axis.


> And you said the clusters are stretching in dimension 2, which could mean the length is working. Is there a way to verify this? Like could we set a colormap and see if there is a gradual increase in color across dimension 2?

Following the previous response, I thought of a way to check that the conditioning works without latent space visualization. It consists of 2 points:
1. Checking that the predicted conditioning variable matches the input. In other words, measuring the MSE between the "true age" VS the "predicted age", their correlation, and plotting one against the other. This way you know how good the encoder is capturing the conditioning variable variation in the training data. The authors do it like this: 
 

2. Visual checking fiber generation for specific bundles. Knowing that different bundles have different fiber lengths, we try to generate fibers of specific length, and see whether the generated fibers belong to the desired bundle (no matter if they are plausible or implausible). Having length as the conditioning variable allows us to perform this trick, what would not be so intuitive to check if we had used Fractional Anisotropy or other DTI-derived metrics.


What do you think? We can discuss these strategies in the weekly meeting today.

**Comment by jhlegarreta** [comment]
Couple of other references where slightly different approaches related to what we discussed:
- https://github.com/mingukkang/Adversarial-AutoEncoder (see the latent space plots there, section 2) : an implementation of: https://arxiv.org/pdf/1511.05644
- https://arxiv.org/pdf/1706.00409: there are probably implementations out there. Hopefully one of them works.

**Comment by pjsjongsung** [comment]
I've looked at the paper and a simple summary would be

After we get the equation for the VAE, the difference between a normal VAE and the model is that now we have a condition c. They omit any proof on deriving the equation, but this adds a KL divergence term between the latent space created conditioned on the input and the latent space created conditioned on the condition(age). Thus, they are creating this latent space conditioned on age by adding a latent generator, the scalar to latent space network, to calculate the KL divergence between the two latent spaces.

I still think there could be too little variability when creating the whole latent space with a single scalar, but their reasoning behind using this framework looks sound. If we later find that there needs to be more variability during inference, we could try to make the latent generator a cGAN instead.

**Comment by itellaetxe** [comment]
@pjsjongsung @jhlegarreta 

I know that the conditional (regression) VAE was left behind, but I forgot to plot the latent space of the last model that I trained, with the **really** good label loss results (reconstruction loss was a bit worse). Just for the record I am including the results here, even if they will be included in the blogs. See the plots below:

```
Mean squared error: 50.371 (MAE = ~ 7)
R2 Variance score: 0.941
```

This plot shows the predicted attribute vs the real value of the attribute. It is interesting to see that almost always the model underestimates the attribute. (we commented on this plot in the last meeting)


And here is the latent space representation of the plausible fibers found in the training dataset.


Interesting! The bundles are grouped into "worm" clusters. Again, there are 7 bundles and 7 worms, and each worm has a constant length, indicating that yes, each bundle is represented by a worm.

(p.d.: Sorry for the word choice but I found this clusters kind of funny-looking.)

**Comment by jhlegarreta** [comment]
Thanks for the report @itellaetxe .

> The bundles are grouped into "worm" clusters. Again, there are 7 bundles and 7 worms

Not sure if I seen the 7, but OK.

**Comment by itellaetxe** [comment]
Merging into `main` to have all the latest code visible in the repo landing page.

Will open the `dev/experimental` branch again for the next experiments.
